### PR TITLE
John conroy/update tables HMP-357 HMP-377 HMP-468 HMP-476

### PR DESCRIPTION
**Pull request by john-conroy**
Create a reusable table component for search API entities. Further updates and use in workspaces dialog in #3313. Also reverts the overwrite of the `action.active` default MUI palette color.

Screenshot 2023-11-08 at 10 07 31 AM
Screenshot 2023-11-08 at 10 07 54 AM

### CODE REVIEW

File: context/app/static/js/shared-styles/tables/EntitiesTable/EntityTable.tsx [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                    </TableCell>
+                  ))}
+                </TableRow>
+              );
+            }
+            return null;
+          })}
+          {tableBodyPadding.bottom > 0 && (
+            <tr>
+              <td style={{ height: `${tableBodyPadding.bottom}px` }} aria-hidden="true" />
```
Comment (NickAkhmetov) [inline context/app/static/js/shared-styles/tables/EntitiesTable/EntityTable.tsx:148]: Given this block has a lot in common with the block on lines 125-129, should we consider adding a `TablePadding` component to create an empty row like this?

```jsx
function TablePadding({ height }: { height?: number }) {
  if (!height || height <== 0) {
    return null;
  } 
  return (
    <tr aria-hidden="true">
      <Box component="td" height={height} />
    </tr>
  )
}
```

File: context/app/static/js/shared-styles/tables/EntitiesTable/EntityTable.tsx [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                  {isSelectable && <SelectableRowCell rowKey={hit?._id} />}
+                  {columns.map(({ cellContent: CellContent, id }) => (
+                    <TableCell key={id}>
+                      <CellContent hit={hit._source} />
+                    </TableCell>
+                  ))}
+                </TableRow>
+              );
+            }
+            return null;
```
Comment (NickAkhmetov) [inline context/app/static/js/shared-styles/tables/EntitiesTable/EntityTable.tsx:144]: I wonder if it's possible/worthwhile to turn this into a component as well?

File: context/app/static/js/shared-styles/tables/EntitiesTable/EntitiesTables.tsx [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  isSelectable?: boolean;
+  initialTabIndex?: number;
+  entities: EntitiesTabTypes<Doc>[];
+}
+
+const singleTabProps = {
+  'aria-disabled': true,
+  disableRipple: true,
+  sx: { cursor: 'default' },
+};
```
Comment (NickAkhmetov) [inline context/app/static/js/shared-styles/tables/EntitiesTable/EntitiesTables.tsx:20]: Does this offer advantages over using `disabled={entities.length === 1}`? https://mui.com/material-ui/react-tabs/#disabled-tab

File: context/app/static/js/shared-styles/tabs/Tabs/Tabs.tsx
Comment (NickAkhmetov) [inline context/app/static/js/shared-styles/tabs/Tabs/Tabs.tsx:1]: Perhaps we could save this for another PR, but it looks like there is an experimental `mui/lab` version of the tabs component which handles ARIA for us: https://mui.com/material-ui/react-tabs/#experimental-api

File: context/app/static/js/shared-styles/tables/EntitiesTable/EntitiesTables.tsx
Comment (NickAkhmetov) [inline context/app/static/js/shared-styles/tables/EntitiesTable/EntitiesTables.tsx:1]: Very minor nitpick:

- Directory is named `EntitiesTable`
- Single entity table component is named `EntityTable`
- Multi-entity table component is named `EntitiesTables`

My suggestion would be to standardize naming around `EntityTable`, naming both the directory and single entity table component `EntityTable` and calling this one `TabbedEntityTable`.

File: context/app/static/js/shared-styles/tabs/Tabs/Tabs.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      sx={{
+        color: 'primaryContainer.contrastText',
+        backgroundColor: 'primaryContainer.main',
+        borderRadius: '4px 4px 0px 0px',
+        flex: 'none',
+      }}
+      {...props}
+    >
+      {children}
+    </MUITabs>
```
Comment (NickAkhmetov) [inline context/app/static/js/shared-styles/tabs/Tabs/Tabs.tsx:18]: This is entirely optional - a neat alternative approach:

```jsx
const Tabs = styled((props: TabsProps) => <Tabs variant="fullWidth" {...props} />)(
	( { theme: { palette: { primaryContainer, action }, spacing } ) => ({
	  color: primaryContainer.contrastText,
	  backgroundColor: primaryContainer.main,
	  borderRadius: spacing(0.5, 0.5, 0, 0),
	  flex: 'none',
	  '& .MuiTabs-indicator': {
	    backgroundColor: action.activeTab,
	    height: spacing(0.5),
	  }
	})
);
```

**Review by NickAkhmetov — APPROVED** [review]
Great work! Some minor suggestions, but nothing worth blocking further development over. Pretty much all my suggestions have to do with further simplification of complex components and minor style improvements, so they could be tackled in a subsequent improvement ticket.

File: context/app/static/js/shared-styles/tables/EntitiesTable/EntitiesTables.tsx (hunk 3 above)
Comment (john-conroy) [reply context/app/static/js/shared-styles/tables/EntitiesTable/EntitiesTables.tsx:20]: Styling. The MUI disabled state is grayed out.

File: context/app/static/js/shared-styles/tabs/Tabs/Tabs.tsx
Comment (john-conroy) [reply context/app/static/js/shared-styles/tabs/Tabs/Tabs.tsx:1]: I saw that and thought the same. I'll make a ticket to revisit.

File: context/app/static/js/shared-styles/tables/EntitiesTable/EntitiesTables.tsx
Comment (john-conroy) [reply context/app/static/js/shared-styles/tables/EntitiesTable/EntitiesTables.tsx:1]: When we adopt the table throughout the repo we can reorganize?

File: context/app/static/js/shared-styles/tables/EntitiesTable/EntityTable.tsx [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                sx={({ palette }) => ({ backgroundColor: palette.background.paper })}
+              />
+            )}
+            {columns.map((column) => (
+              <EntityHeaderCell column={column} setSort={setSort} sortState={sortState} key={column.id} />
+            ))}
+          </TableRow>
+          <TableRow aria-hidden="true">
+            <TableCell
+              colSpan={columns.length + 1}
```
[the remark this answers is not in the document]
Comment (john-conroy) [reply context/app/static/js/shared-styles/tables/EntitiesTable/EntityTable.tsx:108]: Nice catch!

File: context/app/static/js/shared-styles/tables/EntitiesTable/EntityTable.tsx (hunk 2 above)
Comment (john-conroy) [reply context/app/static/js/shared-styles/tables/EntitiesTable/EntityTable.tsx:144]: Not sure if it's worthwhile given the component is only used here. Adds a bit of overhead to pull into a component and repeat types for the `virtualRow`, `searchHits`, and columns. What do you think?

**Review by NickAkhmetov — APPROVED** [review]
Looks great! 👍🏻

**Pull request merged**